### DISCUSSION TITLE
Queue and RAM usage

### DISCUSSION THREAD

**Discussion by RobertMorton56** [earlephilhower/arduino-pico#1401] [Q&A] [answered]
Hi folks,

Really enjoying this core, but I'm just wondering about something. I'm using the pico util queue and compiling the following code with arduino-cli.

```
#define QUEUE_SIZE 10

queue_t queue;
typedef struct queueItem {
	uint8_t message [ 50 ];
	uint8_t length;
} message;

void setup( void ) { queue_init( &queue, sizeof( message ), QUEUE_SIZE ); }

void loop( void ) {}
```

Both QUEUE_SIZE of 5000 and QUEUE_SIZE of 10 show the following after successful compilation:

```
Sketch uses 52220 bytes (2%) of program storage space. Maximum is 2093056 bytes.
Global variables use 8928 bytes (3%) of dynamic memory, leaving 253216 bytes for local variables. Maximum is 262144 bytes.
```

I'd imagine the queue init would allocate memory (which it does: q->data = (uint8_t *)calloc(element_count + 1, element_size); ). Is this allocated memory not reflected in the free memory shown after compilation? Is this an arduino-cli bug, perhaps it's being optimized away?

**Comment by earlephilhower** [comment, marked as answer]
The report shows static allocations for global/statics, only.

The `calloc()` comes from the heap, which is runtime, and is not possible to report w/o actually, err, running the code.

**Comment by ruifig** [comment]
@RobertMorton56 , like earle mentioned, that build report will give you only global/statics. 
To get a report of memory allocated at runtime, you can call `rp2040.getTotalHeap()` and `rp2040.getUsedHeap()`.
Of course that will only give you the amount allocated at the time of the call.
See https://arduino-pico.readthedocs.io/en/latest/rp2040.html

Try adding a `rp2040.getUsedHeap()` at the end of the setup() and see the difference for when you use a queue size of 5000 and 10.
